### ISSUE TITLE
Support making static copy of entire website

### ISSUE THREAD

**Issue by marklagendijk** [s0ph1e/node-website-scraper#43]
I think an important use case of this module would be to make a static copy of an entire website. For this use case the following functionality would be needed:
- [x] The scaper is able to download the HTML of a page.
- [x] The scaper is able to download resources which are used on a page.
- [x] The scaper is able to crawl the entire website.
- [ ] The scaper can be configured to ignore external websites.
- [ ] The scaper can save the files in the same structure as the paths on the website.

E.g:
- / => index.html 
- /css/layout.css => css/layout.css
- /about => about/index.html

**The scaper can be configured to ignore external websites**
This can be implemented by creating an 'urlFilter' option which takes a function. Before a request to a url is done, this function is called with the url (and possible other info which could be useful). The url is only called if the urFilter returns `true`.

**The scaper can save the files in the same structure as the paths on the website**
This can be implemented by creating a 'outputPathGenerator' (or a better name) option which takes a function. If the user specifies this option the output path is generated by passing the `resource` to this function.

**Example**
The following is an example of how you could make a scraper which saves an entire website, using the above described functionality.
``` javascript
var _ = require('lodash');
var pathUtil = require('path');
var urlUtil = require('url');
var websiteUrl = 'http://nodejs.org/'
var options =  {
  urls: [websiteUrl],
  directory: '/path/to/save/',
  recursive: true,
  maxDepth: 10,
  urlFilter: function(url){
    return _.startsWith(url, websiteUrl);
  },
  outputPathGenerator: function(resource, directory){
    var urlObject = urlUtil.parse(resource.url);
    // Todo: add logic to check if it is an HTML page which does not end in '.html'
    // If so, append '/index.html'
    var relativePath = '.' + urlObject.path; 
    return pathUtil.resolve(directory, relativePath);
  }
};
```

**Comment by s0ph1e** [comment]
Hi @marklagendijk 
Thanks for suggestions

As for **The scaper can be configured to ignore external websites** 
I think it is great idea and it is quite simple to implement. I'm going to add it to release `1.0.0`

As for **The scaper can save the files in the same structure as the paths on the website**
I really like the idea and I thought about it earlier
But it is more tricky
For example: user specified `directory` for downloading `/home/user/mysite` and there is url like `../../../etc/someResource.html`. It means that resource should be loaded to `/etc/someResource.html` 
I see next problems here:
* File `/etc/someResource.html` may exist
* User may not have access to `/etc`
* Module may save files everywhere in OS (if it has permissions). It is dangerous

Now `website-scraper` checks that `directory` doesn't exist and creates it. It is responsible only for `directory`. Allowing it create files everywhere is dangerous. Also it requires check existence of each file (and deciding what to do if it exists).

**Comment by marklagendijk** [comment]
@s0ph1e will you accept pull requests for the two features which I described (urlFilter and outputPathGenerator)? I think these features would make the module very flexible, and make it support not only the use cases I described, but also lots of other ones.

The security issue you thought of, does not exist. Any path of an url you get will start with `/`. Also, even if it would exist, you could easily protect against it:

``` javascript
var urlObject = urlUtil.parse(resource.url);
var relativePath = '.' + urlObject.path; 
var absolutePath = pathUtil.resolve(directory, relativePath);
if(!_.startsWith(absolutePath, directory){
  throw new Error('Invalid relative path');
}
```

**Comment by s0ph1e** [comment]
I'll accept PR for *urlFilter* but not for *outputPathGenerator*

The main goal of second suggestion (*outputPathGenerator*) is **full copy** of existing website. 
From my point of view it is impossible to implement.

> Any path of an url you get will start with /

This is true only in ideal world. Let's name sites with contain only such urls _ideal_sites_. In real life sites may contain everything (as in my example above)

> you could easily protect against it:

Yes. But in this case we will not receive **full copy** of site. Instead of it user will receive exception. It means that scraping process will be failed and nothing will be saved. What if user doesn't want this behavior? Maybe he wants to download that resources to another place or skip them? Do you think it is good idea to throw exception where there are 100500 _ideal_resources_ on website and only 1 `../../someResouce.html`? 
In any of this cases we will not receive **full copy** - which is the main goal. We can get **partial copy** of site (where not _ideal_resources_ are skipped or saved to another place etc.) but I don't see advantages over current implementation

That's why I don't want to implement it in this module.

--------------------------

You are welcome to create PR for *urlFilter* and don't forget to write tests for it and update Readme

**Comment by marklagendijk** [comment]
I expect that the cases with urls like `http://my-website.com/../../someResource.html` are already taken care of by one of the used modules. For example browsers also do this. Such url should be normalized to `http://my-website.com/someResource.html`.

If this is not the case it is easy to do this normalization ourselves. The rule is simple: `../` removes the previous non `../` part, unless you are already at the root.

However, I feel that even if there would be an unsolvable issue with this, it would still be a great addition to support custom `generateFilename` functions. In this case the user of the module can write his own `genereateFilename` function, and have full control of the output structure (and solve any issues according to his use case).

**Comment by marklagendijk** [comment]
I created an example of such normalize function. It is called `getSaveRelativePath` and it converts a path of an url to a save relative path.

``` javascript
var path = require('path');

function getSaveRelativePath(urlPath){
    var relativePath = path.normalize('.' + urlPath);
    var saveRelativePath = relativePath.replace(/\.\./g, '.');
    return './' + path.normalize(saveRelativePath);
}

console.log(getSaveRelativePath('/example/test/../../../../../etc/passwd'));
// => ./etc/passwd
```

**Comment by marklagendijk** [comment]
This is my current `file-name-generators/by-site-structure.js`:
``` javascript
var _ = require('underscore');
var url = require('url');
var path = require('path');
var resourceTypes = require('../config/resource-types');

module.exports = function generateFilename (resource, options, loadedResources) {
    var urlObject = url.parse(resource.url);
    if(resource.getType() === resourceTypes.html && !_.contains(resource.filename, '.html')){
        var fileName = options.defaultFilename + (urlObject.search || '');
        return getSaveRelativePath(path.join(urlObject.pathname, fileName));
    }
    else{
        return getSaveRelativePath(urlObject.pathname);
    }
};


function getSaveRelativePath(urlPath){
    var relativePath = path.normalize('.' + urlPath);
    var saveRelativePath = relativePath.replace(/\.\./g, '.');
    return './' + path.normalize(saveRelativePath);
}
```

**Comment by s0ph1e** [comment]
Maybe I didn't understand something.

Could use please explain me next use case:
We want to download `http://my-website.com/path/to/page.html` to directory `/home/user/mypage`. Imagine this page contains resource `../../resource.png` (which resolves to `http://my-website.com/resource.png`). Where should it be saved?

**Comment by marklagendijk** [comment]
Aah, that's a different use case.
My use case is that the whole site is saved to the `/home/user/mywebsite` directory. So:

- `http://my-website.com/path/to/page.html` =>  `/home/user/mywebsite/path/to/page.html`
- `http://my-website.com/path/to/../../resource.png` => `http://my-website.com/resource.png` => `/home/user/mywebsite/resource.png`

**Comment by s0ph1e** [comment]
So your suggestion is only for root - `http://my-website.com/` ?

**Comment by marklagendijk** [comment]
This specific `filenameGenerator` saves the files in the exact same structure as on the site itself. That indeed means that it contains the whole directory structure, relative to the root of the website.

But since it becomes possible to specify your own `filenameGenerator`, the users can write a custom `filenameGenerator` to get any output structure they want.

**Comment by marklagendijk** [comment]
@s0ph1e in 8d00844f5d579ee2ae7f8d2d97682010ad3c2516 I implemented the functionality which I described here. It still lacks tests and documentation.

I tested it with an average Wordpress site and it works perfectly:

1. The `urlFilter` makes sure only urls from the site itself are scraped.
2. The `by-site-structure` `filenameGenerator` converts all the file names into absolute paths on the website in a safe way, e.g: `../../../images/resource.png` on `/some/page/page.html` => `/images/resource.png`.

Could you review the code which I wrote so far and see if you would be ok with merging this in (as soon as the tests and documentation are finished)?

**Comment by marklagendijk** [comment]
Example usage:
``` javascript
var path = require('path');
var siteUrl = 'http://mywebsite.com';
scraper.scrape({
    urls: [ siteUrl ],
    directory: path.resolve(__dirname, 'output/mywebsite.com'),
    recursive: true,
    maxDepth: 100,
    omitDefaultFilenameInUrls: true,
    filenameGenerator: 'bySiteStructure',
    urlFilter: function(url){
        return url.indexOf(siteUrl) === 0;
    }
});
```

**Comment by s0ph1e** [comment]
Thanks @marklagendijk 
I'll review it when I have some free time

**Comment by s0ph1e** [comment]
I reviewed the changes and I'm ok to merge once they will be covered with tests and documentation 👍 

I think need to create separate PRs for `urlFilter` and `filenameGenerators` - it would be easier to review, discuss and merge them
The same for `omitDefaultFilenameInUrls ` (need to discuss name for this feature because it is not clear from my point of view. What about `prettifyUrls`?)

**Comment by marklagendijk** [comment]
Great, I'll do that! (Y)
I'm fine with `prettifyUrls`, it is indeed more intuitive.

**Comment by s0ph1e** [comment]
`urlFilter` implemented in #47
`filenameGenerator` in #51 
`prettifyUrl` in #45
